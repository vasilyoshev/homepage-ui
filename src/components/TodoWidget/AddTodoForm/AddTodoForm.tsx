import React, { useState, ChangeEvent, FormEvent } from 'react';
import { FormGroup, Button, TextField } from '@material-ui/core';
import { useDispatch } from 'react-redux';
import { addTodo } from 'slices';

export const AddTodoForm: React.FC = () => {
  const dispatch = useDispatch();
  const [newTodo, setNewTodo] = useState<string>('');

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setNewTodo(e.target.value);
  };

  const submitTodoItem = (e: FormEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if(!newTodo.trim()) {
      return;
    }
    dispatch(addTodo(newTodo));
    setNewTodo('');
  };

  return (
    <FormGroup>
      <TextField
        type='text'
        label="add new task"
        value={newTodo}
        color="primary"
        margin='normal'
        variant="outlined"
        size="small"
        onChange={handleInputChange}
      />
      <Button
        type="submit"
        size="medium"
        variant="contained"
        color="primary"
        onClick={submitTodoItem}
      >
        Add Todo
      </Button>
    </FormGroup>
  );
};
