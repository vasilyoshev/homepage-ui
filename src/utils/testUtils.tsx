import React from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';

// TODO extract mocks to separate files, e.g. widgetsMock.js
const mockState = {
  loading: {},
  widgets: { widgetsProps: [] },
  user: {},
};
const createMockStore = configureStore([thunk]);
const store = createMockStore(mockState);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Providers = ({ children }: any) => (
  <Provider store={store}>
    <MemoryRouter initialEntries={['/']}>{children}</MemoryRouter>
  </Provider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
  render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';

export { customRender as render };
