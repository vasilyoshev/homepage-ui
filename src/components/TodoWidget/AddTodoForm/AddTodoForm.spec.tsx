import { AddTodoForm } from './AddTodoForm';
import { render, screen } from 'utils';
import userEvent from '@testing-library/user-event';

describe('AddTodoForm', () => {
  it('renders successfully', () => {
    const { asFragment } = render(<AddTodoForm />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should render empty input field', () => {
    render(<AddTodoForm addTodo={jest.fn()} />);
    const input = screen.getByRole('textbox');
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute('type', 'text');
    expect(input).toBeEmptyDOMElement();
  });
  it('should render button with name add todo', () => {
    render(<AddTodoForm addTodo={jest.fn()} />);
    const btn = screen.getByRole('button', {name: /add todo/i});
    expect(btn).toBeInTheDocument();
  });
  it('user can add some text', () => {
    render(<AddTodoForm addTodo={jest.fn()} />);
    const input = screen.getByRole('textbox');
    const btn = screen.getByRole('button', {name: /add todo/i});
    userEvent.click(btn);
    userEvent.type(input, 'todo task');
    expect(input).toHaveValue('todo task');
    expect(input).toBeEmptyDOMElement();
  });
});
