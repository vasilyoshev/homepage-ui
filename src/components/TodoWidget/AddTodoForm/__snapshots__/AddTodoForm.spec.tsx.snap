// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoForm renders successfully 1`] = `
<DocumentFragment>
  <form
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="mui-1"
        id="mui-1-label"
      >
        What needs to be done?
        <span
          aria-hidden="true"
          class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="mui-1"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1in441m"
          >
            <span>
              What needs to be done? *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddCircleOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </span>
      Add Todo
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
